[ios][precompile][spm] make emc work with swift package manager

chrfalch opened the pull request:
# What?

This is the first of a series of PRs to adjust ExpoModulesCore to build using the Swift Package Manager.

Multiple PRs will follow adding support for generating, building and packaging EMC through expo-tools.

This PR contains the simplest set of changes for SPM to be able to build EMC and is mainly about trying to not re-expose symbols (headers) from React Native's prebuilt binaries (xcframeworks).

The main problem that this PR tries to solve is to remove exports of React Native headers. This is a bit tricky when we have classes that inherits from a React Native type - then we need to export the React Native types by including the header files since forward declaring types use for inheritance is not possible in Objective-c.

- Deleted the RCTComponentDataSwiftAdapter since it is no longer used in the new architecture
- Removed use of EXHostWrapper
- Removed re-export of RCTHTTPRequestHandler in ExpoModulesCore.h
- Resolved imports of umbrella headers so that they use the correct paths depending on which "mode" we're building in
- Used built-in macro __building_module combined with forward declaration to avoid re-exporting React headers
- Removed reference to Swift types that hasn't been built when building obj-c with SPM (EXFileSystemLegacyUtilities in EXModuleRegistryProvider.m)
- Removed direct import of Swift interface from obj-c (it isn't built yet when we build the obj-c target)

chrfalch committed:
[ios][precompile][spm] make emc work with swift package manager

This is the first of a series of PRs to adjust ExpoModulesCore to build using SPM - ie. through a definition in a Package.swift.

Another PR will add support for generating, building and packaging EMC through expotools.

This PR contains the simplest set of changes for SPM to be able to build EMC and is mainly about trying to not re-expose symbols (headers) from React Native's prebuilt binaries (xcframeworks).

The main problem that this PR tries to solve is to remove exports of React Native headers. This is a bit tricky when we have classes that inherits from a React Native type - then we need to export the React Native types by including the header files since forward declaring types use for inheritance is not possible in Objective-c.

- Deleted the RCTComponentDataSwiftAdapter since it is no longer used in the new architecture
- Removed use of EXHostWrapper
- Removed re-export of RCTHTTPRequestHandler in ExpoModulesCore.h
- Resolved imports of umbrella headers so that they use the correct paths depending on which "mode" we're building in
- Used built-in macro __building_module combined with forward declaration to avoid re-exporting React headers
- Removed reference to Swift types that hasn't been built when building obj-c with SPM (EXFileSystemLegacyUtilities in EXModuleRegistryProvider.m)
- Removed direct import of Swift interface from obj-c (it isn't built yet when we build the obj-c target)

tsapeta reviewed (approved):
Looks good to me! I think I can help here a bit and entirely remove a few of these files 👍

chrfalch committed:
merge: updated missing imports after merge

chrfalch committed:
Resolving merge issues

1shaked committed:
[docs] Refactor locale assignment with optional chaining (#42690)

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

tsapeta committed:
[iOS][core] Migrate some unit tests to Swift Testing (#42662)

Wenszel committed:
[ci] Update macOS workflow to ignore Android files (#42686)

# Why
There's no need to run the macOS workflow for android changes.

# How
Add an exclusion to paths based on this:

https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#example-including-and-excluding-paths

# Test Plan
Verified in a stacked PR.
